pkg/pbq/pbqmanager.go

```diff
@@ -0,0 +1,204 @@
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/pbq/store"
+	"github.com/numaproj/numaflow/pkg/pbq/store/memory"
+	"github.com/numaproj/numaflow/pkg/pbq/store/noop"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var NonExistentPBQErr error = errors.New("missing PBQ for the partition")
+
+// Manager helps in managing the lifecycle of PBQ instances
+type Manager struct {
+	storeOptions *store.StoreOptions
+	pbqOptions   *options
+	pbqMap       map[string]*PBQ
+	log          *zap.SugaredLogger
+	// we need lock to access pbqMap, since deregister will be called inside pbq
+	// and each pbq will be inside a go routine, and also entire PBQ could be managed
+	// through a go routine (depends on the orchestrator)
+	sync.RWMutex
+}
+
+// NewManager returns new instance of manager
+// We don't intend this to be called by multiple routines.
+func NewManager(ctx context.Context, opts ...PBQOption) (*Manager, error) {
+	pbqOpts := DefaultOptions()
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt(pbqOpts); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	pbqManager := &Manager{
+		pbqMap:       make(map[string]*PBQ),
+		pbqOptions:   pbqOpts,
+		storeOptions: pbqOpts.storeOptions,
+		log:          logging.FromContext(ctx),
+	}
+
+	return pbqManager, nil
+}
+
+// CreateNewPBQ creates new pbq for a partition
+func (m *Manager) CreateNewPBQ(ctx context.Context, partitionID string) (ReadWriteCloser, error) {
+
+	var persistentStore store.Store
+	var err error
+
+	switch m.storeOptions.PbqStoreType() {
+	case dfv1.NoOpType:
+		persistentStore, _ = noop.NewPBQNoOpStore()
+	case dfv1.InMemoryType:
+		persistentStore, err = memory.NewMemoryStore(ctx, partitionID, m.storeOptions)
+		if err != nil {
+			m.log.Errorw("Error while creating persistent store", zap.Any("partitionID", partitionID), zap.Any("storeType", m.storeOptions.PbqStoreType()), zap.Error(err))
+			return nil, err
+		}
+	case dfv1.FileSystemType:
+		return nil, errors.New("not implemented")
+	}
+
+	// output channel is buffered to support bulk reads
+	p := &PBQ{
+		store:       persistentStore,
+		output:      make(chan *isb.Message, m.pbqOptions.channelBufferSize),
+		cob:         false,
+		partitionID: partitionID,
+		options:     m.pbqOptions,
+		manager:     m,
+		log:         logging.FromContext(ctx).With("PBQ", partitionID),
+	}
+
+	m.register(partitionID, p)
+	return p, nil
+}
+
+// ListPartitions returns all the pbq instances
+func (m *Manager) ListPartitions() []*PBQ {
+	m.RLock()
+	defer m.RUnlock()
+
+	pbqList := make([]*PBQ, len(m.pbqMap))
+	i := 0
+	for _, val := range m.pbqMap {
+		pbqList[i] = val
+		i++
+	}
+
+	return pbqList
+}
+
+// GetPBQ returns pbq for the given partitionID
+func (m *Manager) GetPBQ(partitionID string) (ReadWriteCloser, error) {
```

Comment:
Minor - think this could simply return `ReadWriteCloser` without using an error, which would be easier to use - there's no other error could be generated in the implementation, so using `nil` instead of a particular error to tell if there's a PBQ existing is more convenient.

Reply:
Okay